### PR TITLE
[MM-19792] Implement label search select box

### LINKED ISSUE #394: [MM-19792] Implement label search select box
#### Summary

The user should be able to use server-side autocomplete search for labels.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-19792

#### Additional Info

https://community.atlassian.com/t5/Answers-Developer-Questions/How-do-I-use-the-JIRA-API-to-get-a-list-of-label-options-for-a/qaq-p/537723

In the CreateIssueMetadata, “Labels” come with an “autoCompleteUrl” like this:

`https://mmtest.atlassian.net/rest/api/1.0/labels/suggest?query=`

`https://mmtest.atlassian.net/rest/api/1.0/labels/suggest?customFieldId=10071&query=` for custom label fields

Here’s an example: https://github.com/mattermost/mattermost-plugin-jira/blob/59ded5f9c21a6fbcb2088dc9c4fd39a73751cc1d/webapp/src/testdata/cloud-get-create-issue-metadata-for-project-many-fields.json#L138

---
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Plugin: Jira" community channel](https://community.mattermost.com/core/channels/jira-plugin).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/) and our [Plugins Guide](https://developers.mattermost.com/extend/plugins/).

### PR DESCRIPTION
**Pull request by AGMETEOR**
**Summary**
The user should be able to use server-side autocomplete search for labels.

**Ticket**
[https://mattermost.atlassian.net/browse/MM-19792](https://mattermost.atlassian.net/browse/MM-19792)

label-search

Fixes #394

### CODE REVIEW

**Comment by AGMETEOR** [comment]
@hanzei this is ready for review

File: server/client.go [hunk 1]
```diff
@@ -73,6 +74,10 @@ type IssueService interface {
 	UpdateComment(issueKey string, comment *jira.Comment) (*jira.Comment, error)
 }
 
+type LabelService interface {
+	GetLabels(value, url string) (LabelResult, error)
```
Comment (mickmister) [inline server/client.go]: Maybe rename to `SearchLabels`?

File: server/client.go [hunk 2]
```diff
@@ -161,6 +166,27 @@ func (client JiraClient) RESTPostAttachment(issueID string, data []byte, name st
...
 
+func (client JiraClient) GetLabels(value, url string) (LabelResult, error) {
+	apiEndpoint := fmt.Sprintf("%s/rest/api/2/jql/autocompletedata/suggestions", url)
+	req, err := client.Jira.NewRequest("GET", apiEndpoint, nil)
+	if err != nil {
+		return LabelResult{}, err
+	}
+
+	q := req.URL.Query()
+	q.Add("fieldName", "labels")
```
Comment (mickmister) [inline server/client.go]: What are the options for the `fieldName` here? Could we search for a custom label type, or potentially a user/assignee? That would be great if the code is basically identical across all of the autocomplete endpoints.

File: webapp/src/components/jira_label_selector/jira_label_selector.tsx [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    resetInvalidOnChange?: boolean;
+};
+
+type State = {
+    invalid: boolean;
+    error?: string;
+    cachedSelectedOptions: ReactSelectOption[];
+};
+
+export default class JiraLabelSelector extends React.PureComponent<Props, State> {
```
Comment (mickmister) [inline webapp/src/components/jira_label_selector/jira_label_selector.tsx]: @AGMETEOR Do you think there is an opportunity here to have this component be reusable for custom label types, or something like a user search selector? I think the autocomplete endpoints for Jira are very similar.

I'm thinking the frontend can pass to the backend the name of the field it wants to search on, and the backend can construct the url from there.

**Comment by mickmister** [comment]
@jfrerich The epic link selector does have values populated without typing, whereas the label search does not. This is because the epic link selector's `AsyncSelect` component is using `defaultOptions={true}`, which tells the component to fetch upon initial mount of the component. The label search selector is not using this value.

If the Jira autocomplete API supports a blank search (@AGMETEOR would you happen to know this?), then we should be able to support values without typing.

@jfrerich, @asaadmahmood What are your thoughts on this?

**Comment by jfrerich** [comment]
@mickmister, if the criteria you described (`defaultOptions={true}` and the black search supported by the API) is available, then for sure I think we should support showing values without typing!

**Comment by mickmister** [comment]
@DHaussermann We may want to check older versions of Jira Server to verify that this is compatible.

**Comment by mickmister** [comment]
I'm wondering if we share code between other components like the `Epic Link` selector, and possibly the `IssueSearch` select in the `Attach Issue` modal. I believe this would make it easier to allow the `Epic Link` selector and `Label` selector components to be used in the `Create Issue` modal, if they shared view-related code to be displayed in both contexts.

File: go.mod [hunk 4]
```diff
@@ -8,6 +8,7 @@ require (
 	github.com/dghubble/oauth1 v0.5.0
 	github.com/dgrijalva/jwt-go v3.2.0+incompatible
 	github.com/fatih/structs v1.1.0 // indirect
+	github.com/google/go-querystring v1.0.0
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply go.mod]: I have tidied this up.

File: server/client.go [hunk 5]
```diff
@@ -31,6 +31,7 @@ type Client interface {
 	ProjectService
 	SearchService
 	UserService
+	LabelService
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply server/client.go]: I moved the function to the `SearchService` interface

File: server/client.go [hunk 6]
```diff
@@ -161,6 +166,27 @@ func (client JiraClient) RESTPostAttachment(issueID string, data []byte, name st
 	return attachments[0], nil
 }
 
+func (client JiraClient) GetLabels(value, url string) (LabelResult, error) {
+	apiEndpoint := fmt.Sprintf("%s/rest/api/2/jql/autocompletedata/suggestions", url)
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply server/client.go]: I have reused this

File: server/client.go (hunk 1 above)
Comment (AGMETEOR) [reply server/client.go]: I renamed this to something more generic to be used for autocomplete searches of other types of fields

File: server/client.go (hunk 2 above)
Comment (AGMETEOR) [reply server/client.go]: I made this change

File: webapp/src/components/jira_label_selector/jira_label_selector.tsx (hunk 3 above)
Comment (AGMETEOR) [reply webapp/src/components/jira_label_selector/jira_label_selector.tsx]: I have implemented this

File: server/client.go [hunk 7]
```diff
@@ -161,6 +166,27 @@ func (client JiraClient) RESTPostAttachment(issueID string, data []byte, name st
...
+	req, err := client.Jira.NewRequest("GET", apiEndpoint, nil)
+	if err != nil {
+		return LabelResult{}, err
+	}
+
+	q := req.URL.Query()
+	q.Add("fieldName", "labels")
+	q.Add("fieldValue", value)
+	req.URL.RawQuery = q.Encode()
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply server/client.go]: @mickmister got a question on this. I looked at the documentation of the jira API and couldn't find how to do this on the autocomplete end point. Any ideas?

**Comment by AGMETEOR** [comment]
> @jfrerich The epic link selector does have values populated without typing, whereas the label search does not. This is because the epic link selector's `AsyncSelect` component is using `defaultOptions={true}`, which tells the component to fetch upon initial mount of the component. The label search selector is not using this value.
> 
> If the Jira autocomplete API supports a blank search (@AGMETEOR would you happen to know this?), then we should be able to support values without typing.
> 
> @jfrerich, @asaadmahmood What are your thoughts on this?

Yes, this is possible. However, is it still autocomplete if we preload the labels?

File: server/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply server/client.go]: I'm curious how many labels it returns by default. I believe we can use the [CSV import](https://confluence.atlassian.com/adminjiraserver/importing-data-from-csv-938847533.html) to create labels.

From the docs:

> Import issues with multiple labels by:
> - entering each label in a separate column or
> - putting all labels in one column, deliminated by a space

A file with this format should work:
```
Summary, Labels
"The summary", label1 label2 label3...
```

@AGMETEOR If you want to test the autocomplete limits, I would write a small script to generate this CSV file with about 300 labels or so. If you haven't reached the limit at that point try something like 600 or 1100 labels.

**Comment by mickmister** [comment]
> is it still autocomplete if we preload the labels?

I'm thinking that it has a limit to how many it fetches, so it will still be autocomplete. I also can't find any info on the limit in the Jira docs, so we will have to run a test like I describe [here](https://github.com/mattermost/mattermost-plugin-jira/pull/475#discussion_r390031110). Let me know how you feel about this @AGMETEOR

**Comment by AGMETEOR** [comment]
> > is it still autocomplete if we preload the labels?
> 
> I'm thinking that it has a limit to how many it fetches, so it will still be autocomplete. I also can't find any info on the limit in the Jira docs, so we will have to run a test like I describe [here](https://github.com/mattermost/mattermost-plugin-jira/pull/475#discussion_r390031110). Let me know how you feel about this 
@mickmister I am going to try that out and let you know about the results

**Comment by mickmister** [comment]
@AGMETEOR Cool, no rush. Just following up on your comment :slightly_smiling_face:

File: server/client.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (mickmister) [reply server/client.go]: @AGMETEOR The autocomplete endpoints return 15 elements max. I've tested this on Jira 7, Jira 8, and Jira Cloud.

I think we're good with your code here as it's written :+1:

**Comment by AGMETEOR** [comment]
> The code looks good on the backend layer. Please see my reply regarding limits. My only blocking request is to make the select box fetch without typing.
> 
> I've made a followup ticket to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) up the select boxes [here](https://github.com/mattermost/mattermost-plugin-jira/issues/502). I think consolidating the subscribe modal select boxes to be reusable for the `CreateIssueModal` is going to be a significant enhancement. Thanks for you work on the feature @AGMETEOR!

Going to work on it this weekend

**Comment by jasonblais** [comment]
Hey @AGMETEOR - I know it has been a while. Are you interested continuing to work on this pull request? Looks like it's close. cc @jfrerich

**Comment by mickmister** [comment]
@jasonblais I've actually based an open PR on this PR since @AGMETEOR had made progress on some very useful modular code here. The feature set this PR introduces is part of the PR here https://github.com/mattermost/mattermost-plugin-jira/pull/583

**Comment by jasonblais** [comment]
@mickmister thanks! Shall we close this PR in favor of #583?

**Comment by mickmister** [comment]
@jasonblais Yes, I wasn't sure what the process was for this sort of thing. Closing now

@AGMETEOR Thanks for your work on this and all of your contributions to the Jira plugin 🙂

**Pull request closed without merging**